### ISSUE TITLE
TPU session crashes when using JAX

### ISSUE THREAD

**Issue by tdnb** [googlecolab/colabtools#3009]
I am using JAX on a TPU instance and several hours ago it started working very strange. For example, the bellow simple code crashes session on TPU.

```
import jax.tools.colab_tpu
jax.tools.colab_tpu.setup_tpu()

import jax.numpy as jnp
x = jnp.ones(10)
print(x)
```

If I comment out the last line (`print(x)`), it does not crash, or replace with `print(x.shape)` works. Sometimes, the same code does not crash the session but raise the following error.

```
RuntimeError: Backend 'tpu_driver' failed to initialize: DEADLINE_EXCEEDED: Failed to connect to remote server at address: grpc://10.92.96.58:8470. Error from gRPC: Deadline Exceeded. Details: 
```

This is the simplest example but I experienced many another strange behavior on TPU + JAX. I am not very sure if it only happens on my environment but this happen both on Chrome and Firefox.

Are there any changes recently?

**Comment by camierjs** [comment]
I have been encountering the same issue: session crashes.
I've tried folding back to different of `tensorflow`/`jax`/`jaxlib` versions without success so far.
Here is the output log:
![image](https://user-images.githubusercontent.com/20000517/185456791-5acfa2fb-d538-4a06-b54f-9148453468a0.png)

**Comment by cperry-goog** [comment]
Thanks! Tracking internally at b/243038596

**Comment by evanatyourservice** [comment]
I have the same problem as @camierjs using pro+ with TPU backend. I've tried jax version 0.3.14 (default in colab right now), and 0.3.16 with no luck, it crashes as soon as it uses TPU, or tries to copy from TPU to host.

```
import jax.tools.colab_tpu
jax.tools.colab_tpu.setup_tpu()

import jax
jax.devices()

import jax.numpy as jnp
a = jax.device_put(jnp.ones((128, 128)), device=jax.devices()[0])

print(a * 2)
```

This prints the list of TPU devices, creates a, but crashes at print(a * 2).

the log output is:

```
WARNING | 2022-08-19 15:46:37.107623: E external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_driver.cc:265] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
```

Started happening 2 days ago.

**Comment by MatiasCovarrubias** [comment]
I got the same problem. I wrote a comment in the jax discussion: [https://github.com/google/jax/discussions/12011](url). It is a super weird bug. In my case, it is seeing the TPUS, but it crashes when printing a particular type of jax.numpy.arrays.

**Comment by evanatyourservice** [comment]
Any updates on this? I'm still getting the same cuda error as above.

**Comment by MatiasCovarrubias** [comment]
Me too. I am just using GPU.

**Comment by metrizable** [comment]
One work-around that may resolve the issues above is to specify the `tpu_driver_version` on setup to something like `tpu_driver-0.1dev20220801`:

```
jax.tools.colab_tpu.setup_tpu(tpu_driver_version='tpu_driver-0.1dev20220801')
```

**Comment by evanatyourservice** [comment]
Awesome that works!

**Comment by tdnb** [comment]
@metrizable Thank you very much for you advice!

After specifying TPU driver version, no more crush with `print(jnp.ones(10))`, but unfortunately, when I run my previously worked notebook, many another Error happens.

For example, when I run code which contains `jax.lax.all_gather`, it raises `XlaRuntimeError: INVALID_ARGUMENT: Unknown opcode: all-gather`.

Simple example might be:

```
import numpy as np
import jax
import jax.tools.colab_tpu
jax.tools.colab_tpu.setup_tpu(tpu_driver_version='tpu_driver-0.1dev20220801')

x = np.arange(8)
y = jax.pmap(lambda x: jax.lax.all_gather(x, 'i'), axis_name='i')(x)
```
Error
```
UnfilteredStackTrace: jaxlib.xla_extension.XlaRuntimeError: INVALID_ARGUMENT: Unknown opcode: all-gather

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

XlaRuntimeError                           Traceback (most recent call last)
[<ipython-input-1-d5c39f8eea62>](https://localhost:8080/#) in <module>
      5 
      6 x = np.arange(8)
----> 7 y = jax.pmap(lambda x: jax.lax.all_gather(x, 'i'), axis_name='i')(x)

XlaRuntimeError: INVALID_ARGUMENT: Unknown opcode: all-gather
```

I tried older driver `tpu_driver-0.1dev20211031` but did not work as well.

This is not the only problem. It seems many of my previously worked notebook which use JAX on TPU still does not work.

**Comment by evanatyourservice** [comment]
Turns out doesn't work with my code either, getting xla unimplemented error on lax.ftt

**Comment by lukaemon** [comment]
Have the same problem. Driver version trick works so far.

**Comment by patrickvonplaten** [comment]
This is a huge issue actually! We're experiencing this in all our google colabs at Hugging Face at the moment.

Shouldn't this be treated a bit more urgently? 

Google Colab is the no.1 way of **accessing** TPUs and people use JAX because of it's super nice integration with TPUs.
Very few people in the open-source community have access to TPU VMs, but **everybody** has access to Google colabs. Why is JAX treating Google Colab TPUs as 2nd citizens - this doesn't really line up with being a fully open-source library in my opinion.

**Comment by ultrons** [comment]
Hi all,

I’m a product manager for Cloud TPU. On behalf of Cloud TPU, Jax and Colab teams, thank you for surfacing this issue. We ask for your patience as we are working to resolve this and other related issues to unblock JAX TPU users on Colab. There is no excuse for the delay in our response. Support for TPUs on Colab with JAX, PyTorch and TensorFlow is important to us. We will post here as soon as we have an update on this issue.

**Comment by ultrons** [comment]
Update:
This issue has been resolved. Please create a new issue if you experience any new problem. Once again thanks for your inputs.

**Comment by NightMachinery** [comment]
It seems it is necessary to both install the latest versions of JAX manually and supply the TPU driver explicitly:
```
!pip install --upgrade jax jaxlib 

import jax.tools.colab_tpu
jax.tools.colab_tpu.setup_tpu('tpu_driver_20221011')
```

**Comment by xaviergonzalez** [comment]
> jax.tools.colab_tpu.setup_tpu(tpu_driver_version='tpu_driver-0.1dev20220801')

Whe I try the following code in colab
```
import jax
jax.tools.colab_tpu.setup_tpu(tpu_driver_version='tpu_driver-0.1dev20220801')
```

I get the following error message
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
[<ipython-input-6-2704839ece15>](https://localhost:8080/#) in <module>
      1 import jax
----> 2 jax.tools.colab_tpu.setup_tpu(tpu_driver_version='tpu_driver-0.1dev20220801')

AttributeError: module 'jax' has no attribute 'tools'
```
